server/routes/manageConditions/handlers/curfew/curfewRoutes.test.ts

```diff
@@ -0,0 +1,252 @@
+      },
+      {
+        id: 3,
+        field: 'curfewEnd',
+        value: SimpleTime.fromString('05:00 am'),
+        sequence: 2,
+      },
+      {
+        id: 4,
+        field: 'reviewPeriod',
+        value: 'weekly',
+        sequence: 3,
+      },
+    ],
+  }
+
+  let req: Request
+  let res: Response
+
+  beforeEach(() => {
+    req = {
+      params: {
+        licenceId: '1',
+        conditionCode: CURFEW_CONDITION_CODE,
+      },
+      query: {},
+      body: {},
+    } as unknown as Request
+
+    res = {
+      render: jest.fn(),
+      redirect: jest.fn(),
+      status: jest.fn(),
+      locals: {
+        licence: {
+          id: 1,
+          additionalLicenceConditions: [],
+          version: 'version',
+        },
+        user: {
+          username: 'joebloggs',
+        },
+      },
+    } as unknown as Response
+
+    conditionService.getAdditionalConditionByCode.mockResolvedValue(conditionConfig)
+  })
+
+  describe('GET', () => {
+    it('should redirect to additional conditions list page if the additional condition is not found on the licence', async () => {
+      await handler.GET(req, res)
+      expect(res.redirect).toHaveBeenCalledWith('/licence/create/id/1/additional-licence-conditions')
+    })
+
+    it('should redirect to additional conditions list page with fromReviewFlag if the additional condition is not found on the licence', async () => {
+      req.query.fromReview = 'true'
+      await handler.GET(req, res)
+      expect(res.redirect).toHaveBeenCalledWith('/licence/create/id/1/additional-licence-conditions?fromReview=true')
+    })
+
+    it('should format multiple instances of the condition into a single set of form values', async () => {
+      res.locals.licence.additionalLicenceConditions = [curfewInstance1, curfewInstance2] as AdditionalCondition[]
+      const formResponses = {
+        numberOfCurfews: 'Two curfews',
+        curfewStart: SimpleTime.fromString('01:00 am'),
+        curfewEnd: SimpleTime.fromString('02:00 am'),
+        reviewPeriod: 'weekly',
+        numberOfCurfews2: 'Two curfews',
+        curfewStart2: SimpleTime.fromString('04:00 am'),
+        curfewEnd2: SimpleTime.fromString('05:00 am'),
+        reviewPeriod2: 'weekly',
+      }
+
+      await handler.GET(req, res)
+
+      expect(res.render).toHaveBeenCalledWith('pages/manageConditions/curfew/input', {
+        additionalCondition: curfewInstance1,
+        config: conditionConfig,
+        formResponses,
+      })
+    })
+  })
+
+  describe('POST', () => {
+    it('deletes any existing versions of the curfew condtion', async () => {
+      await handler.POST(req, res)
+
+      expect(licenceService.deleteAdditionalConditionsByCode).toHaveBeenCalledWith(
+        CURFEW_CONDITION_CODE,
+        res.locals.licence,
+        res.locals.user
+      )
+    })
+
+    it('adds a number of instances of the curfew condition equal to the user-selection', async () => {
+      jest.spyOn(handler, 'addCurfewCondition')
```

Comment:
This pattern of spying the test under class and verifying its internal implementation is not great practice.
When doing this style of test its better to test the interactions with the unit under test's collaborators. 

Can you verify the licence service was called appropriately?

Reply:
Absolutely I can
I went with this approach because I'm testing `addCurfewCondition` below separately, and by testing calls to `licenceService` from here it could break if `addCurfewCondition` is refactored. That said, if this type of spying is poor practice, I'm happy to change it over.